pip/req/req_uninstall.py

```diff
@@ -131,60 +131,111 @@ def compact(self, paths):
-            if self._refuse:
-                logger.info('Not removing or modifying (outside of prefix):')
-                for path in self.compact(self._refuse):
-                    logger.info(path)
-            if response == 'y':
+            if auto_confirm or self._allowed_to_proceed(verbose):
                 self.save_dir = tempfile.mkdtemp(suffix='-uninstall',
                                                  prefix='pip-')
-                for path in paths:
+
+                for path in sorted(self.compact(self.paths)):
                     new_path = self._stash(path)
                     logger.debug('Removing file or directory %s', path)
                     self._moved_paths.append(path)
                     renames(path, new_path)
                 for pth in self.pth.values():
                     pth.remove()
-                logger.info(
-                    'Successfully uninstalled %s-%s',
-                    self.dist.project_name, self.dist.version
-                )
+
+                logger.info('Successfully uninstalled %s', dist_name_version)
+
+    def _allowed_to_proceed(self, verbose):
+        """Display which files would be deleted and prompt for confirmation
+        """
+
+        def _display(msg, paths):
+            if not paths:
+                return
+
+            logger.info(msg)
+            with indent_log():
+                for path in sorted(self.compact(paths)):
+                    logger.info(path)
+
+        # In verbose mode, display all the files that are going to be deleted.
+        will_remove = list(self.paths)
+        will_skip = set()
+
+        if not verbose:
+            # In non-verbose mode...
+            # Display only the folders that have packages inside them and files
+            # that are not within those folders. Any files in above folders
+            # that would not be deleted would be displayed in a separate
+            # skipped list.
+
+            folders = set()
+            files = set()
+            for path in will_remove:
+                if path.endswith(".pyc"):
+                    continue
+                if path.endswith("__init__.py") or ".dist-info" in path:
+                    folders.add(os.path.dirname(path))
+                files.add(path)
+
+            folders = self.compact(folders)
+
+            # This walks the tree using os.walk to not miss extra folders
+            # that might get added.
+            for folder in folders:
+                for dirpath, _, dirfiles in os.walk(folder):
+                    for fname in dirfiles:
+                        if fname.endswith(".pyc"):
+                            continue
+
+                        file_ = os.path.join(dirpath, fname)
+                        if os.path.isfile(file_) and file_ not in files:
+                            # We are skipping this file. Add it to the set.
+                            will_skip.add(file_)
```

Comment:
I was thinking unittest to test that given a long list of files we do end up with the correct shorter list.